### PR TITLE
MULE-16282: ProcessingStrategies not working when scheduler raises REE - Workaround

### PR DESCRIPTION
**Pull request by mbuchwald**

### CODE REVIEW

File: core-tests/src/test/java/org/mule/runtime/core/internal/processor/strategy/FlowProcessingStrategyTestCase.java [hunk 1]
```diff
@@ -90,7 +90,7 @@ public void fixedProcessingStrategyTakesPrecedenceOverConfig() throws Exception
   public void createDefaultProcessingStrategyIfNoneSpecified() throws Exception {
     flow.initialise();
     assertThat(flow.getProcessingStrategy(),
-               is(instanceOf(new TransactionAwareProactorStreamWorkQueueProcessingStrategyFactory()
+               is(instanceOf(new TransactionAwareProactorStreamEmitterProcessingStrategyFactory()
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/internal/processor/strategy/FlowProcessingStrategyTestCase.java]: let's do this in another commit

File: core/src/main/java/org/mule/runtime/core/internal/construct/DefaultFlowBuilder.java [hunk 2]
```diff
@@ -298,7 +297,7 @@ private void handleOverload(CoreEvent request, Throwable overloadException) {
      */
     @Override
     protected ProcessingStrategyFactory createDefaultProcessingStrategyFactory() {
-      return new TransactionAwareProactorStreamWorkQueueProcessingStrategyFactory();
+      return new TransactionAwareProactorStreamEmitterProcessingStrategyFactory();
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/construct/DefaultFlowBuilder.java]: let's do this in another commit

File: core/src/main/java/org/mule/runtime/core/internal/processor/strategy/ProactorStreamProcessingStrategy.java [hunk 3]
```diff
@@ -210,4 +220,136 @@ public final void dispose() {
       innerSink.dispose();
     }
   }
+
+  private final class RetrySchedulerWrapper implements Scheduler {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/processor/strategy/ProactorStreamProcessingStrategy.java]: isn't this applicable to other processing strategies?

Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/processor/strategy/ProactorStreamProcessingStrategy.java]: implement `ScheduledExecutorService` instead of `Scheduler`

Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/processor/strategy/ProactorStreamProcessingStrategy.java]: move to `org.mule.runtime.core.internal.util.rx`

File: core/src/main/java/org/mule/runtime/core/internal/processor/strategy/ProactorStreamProcessingStrategy.java [hunk 4]
```diff
@@ -210,4 +220,136 @@ public final void dispose() {
...
+    public boolean isTerminated() {
+      return delegate.isTerminated();
+    }
+
+    @Override
+    public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
+      return delegate.awaitTermination(timeout, unit);
+    }
+
+    private <T> Future<T> submit(Supplier<Future<T>> realSubmit) {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/processor/strategy/ProactorStreamProcessingStrategy.java]: rename this method, this is confusing

File: core/src/main/java/org/mule/runtime/core/internal/processor/strategy/ProactorStreamProcessingStrategy.java [hunk 5]
```diff
@@ -210,4 +220,136 @@ public final void dispose() {
...
+
+    @Override
+    public <T> T invokeAny(Collection<? extends Callable<T>> tasks, long timeout, TimeUnit unit)
+        throws InterruptedException, ExecutionException, TimeoutException {
+      return delegate.invokeAny(tasks, timeout, unit);
+    }
+
+    @Override
+    public void execute(Runnable command) {
+      delegate.execute(command);
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/processor/strategy/ProactorStreamProcessingStrategy.java]: add the same logic as in the `submit` methods

File: core/src/main/java/org/mule/runtime/core/internal/processor/strategy/ProactorStreamProcessingStrategy.java [hunk 6]
```diff
@@ -210,4 +220,136 @@ public final void dispose() {
...
+
+    private <T> Future<T> submit(Supplier<Future<T>> realSubmit) {
+      while (!this.isShutdown() && !this.isTerminated()) {
+        try {
+          return realSubmit.get();
+        } catch (RejectedExecutionException ree) {
+          lastRetryTimestamp.set(nanoTime());
+          try {
+            Thread.sleep(SCHEDULER_BUSY_RETRY_INTERVAL_MS);
+          } catch (InterruptedException e) {
```
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/processor/strategy/ProactorStreamProcessingStrategy.java]: Bug Bug: Either re-interrupt this method or rethrow the "InterruptedException". ([squid:S2142](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS2142))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7624&issues=AWh8XXbHotIXnmsUOEzB&open=AWh8XXbHotIXnmsUOEzB)

[ISSUE_KEY]: AWh8XXbHotIXnmsUOEzB

File: core-tests/src/test/java/org/mule/runtime/core/internal/processor/strategy/AbstractProcessingStrategyTestCase.java [hunk 7]
```diff
@@ -720,6 +719,41 @@ public RejectingScheduler(Scheduler delegate) {
...
+      this.delegate = delegate;
+    }
+
+    private Future<?> submit(Supplier<Future<?>> futureSupplier) {
+      while (!delegate.isShutdown() && !delegate.isTerminated()) {
+        try {
+          return futureSupplier.get();
+        } catch (RejectedExecutionException ree) {
+          try {
+            Thread.sleep(5);
```
Comment (sonar-mulesoft) [inline core-tests/src/test/java/org/mule/runtime/core/internal/processor/strategy/AbstractProcessingStrategyTestCase.java]: Code Smell Code Smell: Remove this use of "Thread.sleep()". ([squid:S2925](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS2925))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7624&issues=AWh8XXdPotIXnmsUOEzC&open=AWh8XXdPotIXnmsUOEzC)

[ISSUE_KEY]: AWh8XXdPotIXnmsUOEzC

File: core/src/main/java/org/mule/runtime/core/internal/util/rx/RetrySchedulerWrapper.java [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+import java.util.TimeZone;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.TimeUnit;
+
+public class RetrySchedulerWrapper implements Scheduler {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/util/rx/RetrySchedulerWrapper.java]: javadoc

File: core/src/main/java/org/mule/runtime/core/internal/util/rx/RetrySchedulerWrapper.java [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+  private void runWithRetry(Runnable operation) {
+    while (!this.isShutdown() && !this.isTerminated()) {
+      try {
+        operation.run();
+        return;
+      } catch (RejectedExecutionException ree) {
+        onRetry.run();
+        try {
+          Thread.sleep(retryTime);
+        } catch (InterruptedException e) {
```
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/util/rx/RetrySchedulerWrapper.java]: Bug Bug: Either re-interrupt this method or rethrow the "InterruptedException". ([squid:S2142](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS2142))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7624&issues=AWiAldciotIXnmsUW49q&open=AWiAldciotIXnmsUW49q)

[ISSUE_KEY]: AWiAldciotIXnmsUW49q

**Comment by mbuchwald** [comment]
--verify

**Comment by mbuchwald** [comment]
--validate

**Comment by mbuchwald** [comment]
--validate

**Comment by mbuchwald** [comment]
--validate

**Comment by mbuchwald** [comment]
--validate
